pkg/kubelet/dockershim/docker_stats.go

```diff
@@ -17,16 +17,112 @@ limitations under the License.
+// ContainerStats returns stats for a container stats request based on container id.
+func (ds *dockerService) ContainerStats(containerStatsRequest *runtimeapi.ContainerStatsRequest) (*runtimeapi.ContainerStatsResponse, error) {
+	containerStats, err := ds.createContainerStats(containerStatsRequest.ContainerId)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &runtimeapi.ContainerStatsResponse{
+		Stats: containerStats,
+	}, nil
 }
 
-func (ds *dockerService) ListContainerStats(*runtimeapi.ListContainerStatsRequest) (*runtimeapi.ListContainerStatsResponse, error) {
-	return nil, fmt.Errorf("Not implemented")
+// ListContainerStats returns stats for a list container stats request based on a filter.
+func (ds *dockerService) ListContainerStats(containerStatsRequest *runtimeapi.ListContainerStatsRequest) (*runtimeapi.ListContainerStatsResponse, error) {
+	filter := &runtimeapi.ContainerFilter{}
+
+	if containerStatsRequest.Filter != nil {
+		filter.Id = containerStatsRequest.Filter.Id
+		filter.PodSandboxId = containerStatsRequest.Filter.PodSandboxId
+		filter.LabelSelector = containerStatsRequest.Filter.LabelSelector
+	}
+
+	containers, err := ds.ListContainers(filter)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var stats []*runtimeapi.ContainerStats
+	for _, container := range containers {
+		containerStats, err := ds.createContainerStats(container.Id)
+
+		if err != nil {
+			return nil, err
+		}
+
+		stats = append(stats, containerStats)
+	}
+
+	return &runtimeapi.ListContainerStatsResponse{
+		Stats: stats,
+	}, nil
+}
+
+func (ds *dockerService) createContainerStats(containerID string) (*runtimeapi.ContainerStats, error) {
+	statsJSON, err := ds.client.GetContainerStats(containerID)
+
+	if err != nil {
+		return nil, err
+	}
+
+	containerJSON, err := ds.client.InspectContainerWithSize(containerID)
+
+	if err != nil {
+		return nil, err
+	}
+
+	status, err := ds.ContainerStatus(containerID)
+
+	if err != nil {
+		return nil, err
+	}
+
+	dockerStats := statsJSON.Stats
+	containerStats := &runtimeapi.ContainerStats{
+		Attributes: &runtimeapi.ContainerAttributes{
+			Id:          containerID,
+			Metadata:    status.Metadata,
+			Labels:      status.Labels,
+			Annotations: status.Annotations,
+		},
+		Cpu: &runtimeapi.CpuUsage{
+			Timestamp:            time.Now().UnixNano(),
+			UsageCoreNanoSeconds: &runtimeapi.UInt64Value{Value: ds.getCpu(&dockerStats)},
+		},
+		Memory: &runtimeapi.MemoryUsage{
+			Timestamp:       time.Now().UnixNano(),
+			WorkingSetBytes: &runtimeapi.UInt64Value{Value: ds.getMemory(&dockerStats)},
+		},
+		WritableLayer: &runtimeapi.FilesystemUsage{
+			Timestamp: time.Now().UnixNano(),
+			StorageId: &runtimeapi.StorageIdentifier{Uuid: containerJSON.Driver},
+			UsedBytes: &runtimeapi.UInt64Value{Value: uint64(*containerJSON.SizeRw)},
+		},
+	}
+	return containerStats, nil
+}
+
+func (ds *dockerService) getMemory(dockerStats *dockertypes.Stats) uint64 {
+	if runtime.GOOS == "windows" {
```

Comment:
This file should be split into _windows.go and _linux.go, so the if/else will not be needed.

Reply:
Fixed.